### PR TITLE
Add JobMaster's ability to throttle tasks from running in JobWorker

### PR DESCRIPTION
**Pull request by bradyoo**
@calvinjia: Note that this is very different from the implementation we have talked about. This is primarily because:

1. There are other priorities being pushed and therefore, I took a bit of an easier path.
2. This is more extensible because it adds throttling to ExecutorService in general which is a common pattern we use.

Downsides of this implementation:

1. The queue can grow unbounded. In this case, it can grow up to Integer.MAX_VALUE.

2. Becuase the max number of threads of the thread pool executor is set to a very high number,
giving bound to the queue has unintended side effects where the thread pool executor will start launching new threads. I can change the implementation to account for this.

3. The current cancel implementation needs to be improved to where the task gets cancelled even when it's in the queue for the case when it gets popped off the queue. But I also don't think I will have much success removing items from this queue proactively.

In a very foreseeable Future:

Dynamically change this number based on WorkerHealth.

### CODE REVIEW

File: job/common/src/main/java/alluxio/job/wire/JobWorkerHealth.java [hunk 1]
```diff
@@ -35,10 +36,11 @@
    * @param loadAverage output of CentralProcessor.getSystemLoadAverage on the worker
    * @param hostname hostname of the worker
    */
-  public JobWorkerHealth(long workerId, double[] loadAverage, String hostname) {
+  public JobWorkerHealth(long workerId, double[] loadAverage, int taskPoolSize, String hostname) {
```
Comment (gpang) [inline job/common/src/main/java/alluxio/job/wire/JobWorkerHealth.java]: NIT: update the javadoc for this parameter

File: job/common/src/main/java/alluxio/job/wire/JobWorkerHealth.java [hunk 2]
```diff
@@ -84,7 +94,8 @@ public String getHostname() {
    */
   public alluxio.grpc.JobWorkerHealth toProto() {
     alluxio.grpc.JobWorkerHealth.Builder builder = alluxio.grpc.JobWorkerHealth.newBuilder()
-        .setWorkerId(mWorkerId).addAllLoadAverage(mLoadAverage).setLastUpdated(mLastUpdated);
+        .setWorkerId(mWorkerId).addAllLoadAverage(mLoadAverage).setTaskPoolSize(mTaskPoolSize)
```
Comment (gpang) [inline job/common/src/main/java/alluxio/job/wire/JobWorkerHealth.java]: is there a unittest for this method?

File: tests/src/test/java/alluxio/job/master/JobMasterIntegrationTest.java [hunk 3]
```diff
@@ -142,4 +148,14 @@ public void getAllWorkerHealth() throws Exception {
     assertNotNull(workerHealth.getHostname());
     assertEquals(3, workerHealth.getLoadAverage().size());
   }
+
+  @Test
+  public void testThreadPoolSize() throws Exception {
+    mJobMaster.setTaskPoolSize(0);
+    long jobId = mJobMaster.run(new SleepJobConfig(1));
+    CommonUtils.sleepMs(300);
```
Comment (gpang) [inline tests/src/test/java/alluxio/job/master/JobMasterIntegrationTest.java]: why is the sleep necessary?

File: tests/src/test/java/alluxio/job/master/JobMasterIntegrationTest.java [hunk 4]
```diff
@@ -76,6 +78,7 @@ public void after() throws Exception {
     mLocalAlluxioJobCluster.stop();
   }
 
+  @Ignore
```
Comment (gpang) [inline tests/src/test/java/alluxio/job/master/JobMasterIntegrationTest.java]: why are all of these ignored?

File: job/server/src/main/java/alluxio/master/job/command/CommandManager.java [hunk 5]
```diff
@@ -90,8 +88,27 @@ public synchronized void submitCancelTaskCommand(long jobId, long taskId, long w
...
+
+    JobCommand.Builder command = JobCommand.newBuilder();
+    command.setSetTaskPoolSizeCommand(setTaskPoolSizeCommand);
+    submit(workerId, command);
+  }
+
+  private void submit(long workerId, JobCommand.Builder command) {
     if (!mWorkerIdToPendingCommands.containsKey(workerId)) {
-      mWorkerIdToPendingCommands.put(workerId, Lists.<JobCommand>newArrayList());
+      mWorkerIdToPendingCommands.put(workerId, Lists.newArrayList());
```
Comment (gpang) [inline job/server/src/main/java/alluxio/master/job/command/CommandManager.java]: Is this already protected by a lock?

File: job/server/src/main/java/alluxio/worker/job/task/ThrottleableThreadPoolExecutor.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        mThrottleReduced.await();
+      }
+    } catch (InterruptedException e) {
+      t.interrupt();
+    } finally {
+      mThrottleLock.unlock();
+    }
+  }
+
+  private boolean check(Thread t) {
```
Comment (gpang) [inline job/server/src/main/java/alluxio/worker/job/task/ThrottleableThreadPoolExecutor.java]: can we rename this to be more descriptive, maybe like "isRunnable" or if reversed "isThrottled"?

File: job/server/src/main/java/alluxio/worker/job/task/ThrottleableThreadPoolExecutor.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+      if (prevThrottlePercentage > mThrottlePercentage) {
+        mThrottleReduced.signalAll();
+      }
+    } finally {
+      mThrottleLock.unlock();
+    }
+  }
+
+  @Override
+  protected void beforeExecute(Thread t, Runnable r) {
```
Comment (gpang) [inline job/server/src/main/java/alluxio/worker/job/task/ThrottleableThreadPoolExecutor.java]: Does this mean if a thread already started, it will still continue to run, even if the throttle percentage increases?

File: job/server/src/main/java/alluxio/worker/job/task/ThrottleableThreadPoolExecutor.java [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+    super.beforeExecute(t, r);
+    mThrottleLock.lock();
+    try {
+      while (!check(t)) {
+        mThrottleReduced.await();
```
Comment (gpang) [inline job/server/src/main/java/alluxio/worker/job/task/ThrottleableThreadPoolExecutor.java]: could this cause some sort of deadlock where a running threads depends on a throttled thread?

File: job/server/src/main/java/alluxio/worker/job/task/TaskExecutorManager.java [hunk 9]
```diff
@@ -40,7 +42,9 @@
 public class TaskExecutorManager {
   private static final Logger LOG = LoggerFactory.getLogger(TaskExecutorManager.class);
 
-  private final ExecutorService mTaskExecutionService;
+  private static final int MAX_TASK_EXECUTOR_POOL_SIZE = 100;
```
Comment (gpang) [inline job/server/src/main/java/alluxio/worker/job/task/TaskExecutorManager.java]: should this be a configurable parameter? Also, should the default be larger?

File: tests/src/test/java/alluxio/job/master/JobMasterIntegrationTest.java (hunk 4 above)
Comment (bradyoo) [reply tests/src/test/java/alluxio/job/master/JobMasterIntegrationTest.java]: Oops :/

File: tests/src/test/java/alluxio/job/master/JobMasterIntegrationTest.java (hunk 3 above)
Comment (bradyoo) [reply tests/src/test/java/alluxio/job/master/JobMasterIntegrationTest.java]: The sleep is to show that this job never completes despite the waiting.

File: job/common/src/main/java/alluxio/job/wire/JobWorkerHealth.java (hunk 2 above)
Comment (bradyoo) [reply job/common/src/main/java/alluxio/job/wire/JobWorkerHealth.java]: There is just an integration test for this. Let me add a unit test

File: job/common/src/main/java/alluxio/job/wire/JobWorkerHealth.java (hunk 1 above)
Comment (bradyoo) [reply job/common/src/main/java/alluxio/job/wire/JobWorkerHealth.java]: This should have been fixed.

File: job/server/src/main/java/alluxio/worker/job/task/ThrottleableThreadPoolExecutor.java (hunk 6 above)
Comment (bradyoo) [reply job/server/src/main/java/alluxio/worker/job/task/ThrottleableThreadPoolExecutor.java]: This file was meant to be deleted. I used to use this but now I don't.

File: job/server/src/main/java/alluxio/worker/job/task/TaskExecutorManager.java (hunk 9 above)
Comment (bradyoo) [reply job/server/src/main/java/alluxio/worker/job/task/TaskExecutorManager.java]: I can move this to something ridiculously large. There is no need to configure this because this can be very very large. I can move it to like a 1000 so that we wouldn't have to care.

File: job/server/src/main/java/alluxio/master/job/command/CommandManager.java (hunk 5 above)
Comment (bradyoo) [reply job/server/src/main/java/alluxio/master/job/command/CommandManager.java]: Every other public method caller to this is synchronized so since it doesn't make a difference, let me add synchronized to here as well.

**Comment by bradyoo** [comment]
feedback addressed

**Comment by bradyoo** [comment]
I changed the design of this and it's no longer working the way I think it should.

**Comment by bradyoo** [comment]
Addressed some feedback. But more importantly, implementation changed quite a bit.

**Comment by bradyoo** [comment]
changed description and fixed that failed test.

File: job/common/src/main/java/alluxio/job/wire/JobWorkerHealth.java [hunk 10]
```diff
@@ -80,16 +106,55 @@ public String getHostname() {
   }
 
   /**
-   * @return proto representation of JobWorkerInfo
+   * @return proto representatx`ion of JobWorkerInfo
```
Comment (gpang) [inline job/common/src/main/java/alluxio/job/wire/JobWorkerHealth.java]: revert this change

File: tests/src/test/java/alluxio/job/master/JobMasterIntegrationTest.java (hunk 3 above)
Comment (gpang) [reply tests/src/test/java/alluxio/job/master/JobMasterIntegrationTest.java]: Is there a way to not depend on sleeps for testing? I would like for integration tests to complete in less time...

File: job/server/src/main/java/alluxio/worker/job/task/PausableThreadPoolExecutor.java [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * ThreadPoolExecutor that can be temporarily paused to prevent any threads from
+ * executing new tasks.
+ */
+public class PausableThreadPoolExecutor extends ThreadPoolExecutor {
+
+  private boolean mIsPaused;
+  private int mNumPaused;
```
Comment (gpang) [inline job/server/src/main/java/alluxio/worker/job/task/PausableThreadPoolExecutor.java]: can you comment on these variables? Especially, I think access has to be protected by the `mPauseLock`?

File: job/server/src/main/java/alluxio/worker/job/task/PausableThreadPoolExecutor.java [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+  @Override
+  protected void beforeExecute(Thread t, Runnable r) {
+    super.beforeExecute(t, r);
+    mPauseLock.lock();
+    try {
+      mNumPaused++;
+      while (mIsPaused) {
+        mUnpaused.await();
+      }
+      mNumPaused--;
```
Comment (gpang) [inline job/server/src/main/java/alluxio/worker/job/task/PausableThreadPoolExecutor.java]: does this not have to be in the `finally` block?

File: job/server/src/main/java/alluxio/worker/job/task/PausableThreadPoolExecutor.java [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+    mNumPaused = 0;
+    mPauseLock =  new ReentrantLock();
+    mUnpaused = mPauseLock.newCondition();
+  }
+
+  /**
+   * @return number of active threads subtracted by the number of tasks paused
+   */
+  public int getNumActiveTasks() {
+    return super.getActiveCount() - mNumPaused;
```
Comment (gpang) [inline job/server/src/main/java/alluxio/worker/job/task/PausableThreadPoolExecutor.java]: Is it correct that `mPauseLock` is not protecting `mNumPaused`?

Comment (bradyoo) [reply job/server/src/main/java/alluxio/worker/job/task/PausableThreadPoolExecutor.java]: Access to mNumPaused is not protected but that's okay. Because protect it or not, the value of mNumPaused can be inaccurate. But being slightly inaccurate here is okay.

**Comment by bradyoo** [comment]
I took a look at the test times. It appears that the before block of launching is taking longer than the test itself. So it's better for me to spend time merging tests together then it is to try to remove sleeps here.

**Comment by bradyoo** [comment]
Feedback addressed

**Comment by bradyoo** [comment]
Ping?

File: job/server/src/main/java/alluxio/worker/job/task/TaskExecutorManager.java [hunk 14]
```diff
@@ -61,11 +64,43 @@ public TaskExecutorManager(int taskExecutorPoolSize, WorkerNetAddress address) {
...
+   */
+  public int getNumActiveTasks() {
+    return mTaskExecutionService.getNumActiveTasks();
+  }
+
+  /**
+   * @return task executor pool size
+   */
+  public int getTaskExecutorPoolSize() {
+    return mTaskExecutionService.getCorePoolSize();
```
Comment (calvinjia) [inline job/server/src/main/java/alluxio/worker/job/task/TaskExecutorManager.java]: This will always return `taskExecutorPoolSize` (# of core threads), is that what we want?

File: job/server/src/main/java/alluxio/worker/job/task/TaskExecutorManager.java [hunk 15]
```diff
@@ -61,11 +64,43 @@ public TaskExecutorManager(int taskExecutorPoolSize, WorkerNetAddress address) {
...
+
+  /**
+   * @param taskExecutorPoolSize number of threads in the task executor pool
+   */
+  public void setTaskExecutorPoolSize(int taskExecutorPoolSize) {
+    Preconditions.checkArgument(taskExecutorPoolSize >= 0);
+    Preconditions.checkArgument(taskExecutorPoolSize <= MAX_TASK_EXECUTOR_POOL_SIZE);
+
+    if (taskExecutorPoolSize == 0) {
+      // treat 0 as a schedule case because ThreadedTaskExecutorService can't seem to have 0 threads
```
Comment (calvinjia) [inline job/server/src/main/java/alluxio/worker/job/task/TaskExecutorManager.java]: `schedule` -> `special`?

File: job/server/src/main/java/alluxio/worker/job/task/TaskExecutorManager.java [hunk 16]
```diff
@@ -61,11 +64,43 @@ public TaskExecutorManager(int taskExecutorPoolSize, WorkerNetAddress address) {
...
+      mTaskExecutionService.pause();
+    } else {
+      mTaskExecutionService.resume();
```
Comment (calvinjia) [inline job/server/src/main/java/alluxio/worker/job/task/TaskExecutorManager.java]: Why do we need to call resume here?

File: job/server/src/main/java/alluxio/worker/job/task/TaskExecutorManager.java (hunk 14 above)
Comment (bradyoo) [reply job/server/src/main/java/alluxio/worker/job/task/TaskExecutorManager.java]: ?? Doesn't that make sense it's getTaskExecutorPoolSize.

File: job/server/src/main/java/alluxio/worker/job/task/PausableThreadPoolExecutor.java (hunk 13 above)
Comment (gpang) [reply job/server/src/main/java/alluxio/worker/job/task/PausableThreadPoolExecutor.java]: Ok. Does it have to be `volatile` then? I'm not entirely sure how "stale" a non-volatile variable might be.

**Comment by bradyoo** [comment]
addressed feedback

**Comment by bradyoo** [comment]
[ERROR] Increment of volatile field alluxio.worker.job.task.PausableThreadPoolExecutor.mNumPaused in alluxio.worker.job.task.PausableThreadPoolExecutor.beforeExecute(Thread, Runnable) [alluxio.worker.job.task.PausableThreadPoolExecutor] At PausableThreadPoolExecutor.java:[line 96] VO_VOLATILE_INCREMENT
[ERROR] Increment of volatile field alluxio.worker.job.task.PausableThreadPoolExecutor.mNumPaused in alluxio.worker.job.task.PausableThreadPoolExecutor.beforeExecute(Thread, Runnable) [alluxio.worker.job.task.PausableThreadPoolExecutor] At PausableThreadPoolExecutor.java:[line 103] VO_VOLATILE_INCREMENT
[INFO] 

@gpang: You sure this is the right idea? Actually, I'll just use the lock.

**Comment by bradyoo** [comment]
addressed feedback.

**Comment by bradyoo** [comment]
alluxio-bot, test this please

**Comment by bradyoo** [comment]
Can we merge this?

**Comment by gpang** [comment]
alluxio-bot, merge this please
